Rework datasource configuration

gsmet opened the pull request:
/cc @Sanne @geoand @emmanuelbernard 

* the compatibility layer for legacy configuration is coming as a subsequent PR - I started working on it but it's still a work in progress
* it's still using JDBC URLs: we want to support them anyway so I thought I would first implement things keeping them and then work later on supporting a split configuration. Split configuration won't be for this iteration.

I thought it would be better to have the gist of it early.

gsmet committed:
Extract the shared datasource config to a separate extension

gsmet committed:
Make reactive clients share datasource configuration

gsmet committed:
Introduce DatabaseKind in a quarkus-datasource-common extension

gsmet committed:
Upgrade the default PostgreSQL dialect to PostgreSQL 10

gsmet committed:
Resolve automatically the SQL Server dialect

dmlloyd committed:
Fix bug where name conflicts in generated config *GetEnclosing are possible

gsmet committed:
Make driver optional and provide a JDBC driver infrastructure

Co-authored-by: Anamarija Talijanac <anamarija.medic@gmail.com>

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -221,8 +221,8 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
 
-            nativeConfig.additionalBuildArgs.ifPresent(l -> l.stream().map(String::trim).forEach(command::add));
             command.add("--initialize-at-build-time=");
+            nativeConfig.additionalBuildArgs.ifPresent(l -> l.stream().map(String::trim).forEach(command::add));
```
What's the rationale behind this? Not that it's particularly important :)

gsmet replied:
Ah sorry, it shouldn't have been there. I removed it.

geoand replied:
Thanks

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/configuration/matching/FieldContainer.java:
```diff
@@ -41,6 +41,7 @@ StringBuilder getCombinedName(final StringBuilder sb) {
         if (sb.length() > 0) {
             sb.append(':');
         }
+        sb.append(member.getField().getDeclaringClass().getSimpleName()).append('#');
```
I assume this is one has been blessed by @dmlloyd ?

geoand replied:
Nevermind, I just saw the authod :)

geoand commented on extensions/datasource/common/src/main/java/io/quarkus/datasource/common/runtime/DatabaseKind.java:
```diff
@@ -0,0 +1,114 @@
+    private static final Map<String, String> ALIASES;
+
+    static {
+        Map<String, String> aliases = new HashMap<String, String>();
```
Maybe add the initial capacity?

geoand replied:
Also, you can just use the `<>` operator to avoid the extra generic type definition

gsmet replied:
It's not easy to have the size of the map given you can have several aliases. Better leave it as that.

Granted for the diamond operator.

geoand commented on extensions/datasource/common/src/main/java/io/quarkus/datasource/common/runtime/DatabaseKind.java:
```diff
@@ -0,0 +1,114 @@
+    private DatabaseKind() {
+    }
+
+    enum SupportedDatabaseKind {
```
This can be private I think

gsmet replied:
Sure.

geoand commented on extensions/datasource/common/src/main/java/io/quarkus/datasource/common/runtime/DatabaseKind.java:
```diff
@@ -0,0 +1,114 @@
+
+        final String mainName;
+        final Set<String> aliases;
+        static final Map<String, String> ALIASES = new HashMap<String, String>();
```
This doesn't seem to be used

gsmet replied:
Removed.

geoand commented on extensions/datasource/common/src/main/java/io/quarkus/datasource/common/runtime/DatabaseKind.java:
```diff
@@ -0,0 +1,114 @@
+
+        SupportedDatabaseKind(String mainName, String... aliases) {
+            this.mainName = mainName;
+            this.aliases = new HashSet<>();
```
Can we add the capacity here?

gsmet replied:
Computing capacity for something hash based requires some computation. It will make the code less readable and it's really not worth it given it's not a hot path.

geoand replied:
Sure, not a big deal

geoand commented on extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/DataSourceProcessor.java:
```diff
@@ -0,0 +1,5 @@
+package io.quarkus.datasource.deployment;
+
+public class DataSourceProcessor {
```
Is this leftover or something?

gsmet replied:
No. IIRC, the javadoc plugin complains when releasing when an artifact is entirely empty.

geoand replied:
Can we then call it something other than `DataSourceProcessor`? It just seems a little missleading to someone who reads it for the first time. 
Otherwise a tiny bit of Javadoc would be nice

geoand commented on extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DataSourceRuntimeConfig.java:
```diff
@@ -0,0 +1,39 @@
+     * For Vault it is: vault-credentials-provider. Not necessary if there is only one credentials provider available.
+     */
+    @ConfigItem
+    public Optional<String> credentialsProviderType = Optional.empty();
```
I don't think we need to set the value. We don't do it in any of the other configurations

gsmet replied:
We do. Because I also initialize them via constructor. You cannot trigger the initialization of another map given you have a value present in a map. And when I have some build time config in the map, I need a valid runtime config in the other map so I need to initialize it myself. Talked to David about that a while ago but it's not trivial.

We used this pattern for Flyway IIRC.

geoand replied:
Understood, thanks

geoand commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -204,11 +205,16 @@ public void build(RecorderContext recorderContext, HibernateOrmRecorder recorder
             return;
         }
 
+        // we only support the default datasource for now
```
I think we should at least warn users if more than one is configured

gsmet replied:
Why? They can be used for other purposes. And this patch doesn't change that, it just changes how the default is propagated.

geoand replied:
I am just saying that a user has no way of knowing this and as someone who looked at it for the first time, it seems suprising to make such a assumption without any warning or anything.
You know this part way better than almost anyone, so I'll leave it to you to decide

gsmet replied:
It has been like that since forever so let's not discuss this here. We have an issue for multiple PU support and a lot of people complaining there so I think users are aware of that :).

geoand replied:
Fine by my, since I'm not the one that will be listening to those users complain :P

geoand commented on extensions/agroal/spi/src/main/java/io/quarkus/agroal/deployment/JdbcDriverBuildItem.java:
```diff
@@ -0,0 +1,32 @@
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class JdbcDriverBuildItem extends MultiBuildItem {
```
Let's have some minimal Javadoc for this one - it seems like a pretty Build Item :)

gsmet replied:
Done.

geoand commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/JdbcDriver.java:
```diff
@@ -0,0 +1,42 @@
+
+    String name();
+
+    public static class JdbcDriverLiteral extends AnnotationLiteral<io.quarkus.agroal.JdbcDriver>
```
```suggestion
    class JdbcDriverLiteral extends AnnotationLiteral<io.quarkus.agroal.JdbcDriver>
```

gsmet replied:
Done.

geoand commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalConnectionConfigurer.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface AgroalConnectionConfigurer {
+
+    static final Logger log = Logger.getLogger(AgroalConnectionConfigurer.class.getName());
```
```suggestion
    Logger log = Logger.getLogger(AgroalConnectionConfigurer.class.getName());
```

gsmet replied:
Done.

geoand commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java:
```diff
@@ -47,61 +55,52 @@
-        String url = dataSourceRuntimeConfig.url.get();
+        InstanceHandle<AgroalConnectionConfigurer> agroalConnectionConfigurerHandle = Arc.container().instance(
+                AgroalConnectionConfigurer.class,
+                new JdbcDriverLiteral(dataSourceBuildTimeConfig.kind.get()));
```
It seems like this hasn't been preceded by a `isPresent` anywhere.

gsmet replied:
No, if I'm there, I'm sure it has been defined.

geoand replied:
OK

geoand commented on extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/ReactiveDataSourceProcessor.java:
```diff
@@ -0,0 +1,5 @@
+package io.quarkus.reactive.datasource.deployment;
+
+public class ReactiveDataSourceProcessor {
```
Is this empty class needed for anything?

gsmet replied:
See previous comment.

geoand replied:
See previous answer :P

geoand reviewed (commented):
I added a bunch of small comments

geoand commented:
Great job on this!!!!

gsmet committed:
Support legacy configuration for Agroal

gsmet committed:
Support legacy configurations for the reactive clients

Sanne commented:
Is it still straight forward to create a datasource configuration for a database which isn't in the "closed set" of supported kinds?

e.g. how would it look like to configure a Datasource conneting to Oracle?

gsmet commented:
@Sanne yes, it should be supported. I had that use case in mind. I should add a test though. Let me do that.

That's the reason why the DatabaseKind manipulates string. Adding a proper driver just improves the experience.

gsmet commented:
@geoand I addressed your comments. I also just added a commit to add the legacy configuration support for Agroal. Still need to work on the legacy config for the reactive clients though.

emmanuelbernard commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcBuildTimeConfig.java:
```diff
@@ -6,27 +6,39 @@
+     * The datasource driver short name
+     */
+    @ConfigItem
+    public Optional<String> shortName;
```
is that to be used as `....jdbc.shortname = h2` ? along side `kind`? I am not following.

gsmet replied:
Ah, this shouldn't be there, it's a leftover. Let me fix that.

gsmet replied:
Fixed.

emmanuelbernard commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcRuntimeConfig.java:
```diff
@@ -111,17 +82,17 @@
      */
     @ConfigItem(defaultValue = "true")
-    public boolean detectStatementLeaks;
+    public boolean detectStatementLeaks = true;
```
@Sanne you want this to be default = true?

gsmet replied:
It was already defaulted to true (see the ConfigItem) so if we need to discuss this, let's not block this PR.

emmanuelbernard replied:
Sure,

Sanne replied:
yes we want that to be true (as it currently is).  We'll have ORM bypass it, but that's for another day ;)

emmanuelbernard commented on extensions/agroal/deployment/src/test/resources/application-default-datasource.properties:
```diff
@@ -1,16 +1,17 @@
 #tag::basic[]
-quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:default
-quarkus.datasource.driver=org.h2.Driver
+quarkus.datasource.kind=h2
```
I prefer `system` or `db-system` (or maybe `product`) than `kind` see https://groups.google.com/forum/#!searchin/quarkus-dev/kind%7Csort:date/quarkus-dev/zos4bOUVNHY/tYcS7IwCBAAJ

CC @Sanne

gsmet replied:
Can't say like it. @Sanne WDYT?

emmanuelbernard commented on extensions/datasource/common/src/main/java/io/quarkus/datasource/common/runtime/DatabaseKind.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * We don't use an enum as we also need to support non built-in database kinds.
+ */
+public final class DatabaseKind {
```
See comment about "system"

emmanuelbernard commented on extensions/datasource/common/src/main/java/io/quarkus/datasource/common/runtime/DatabaseKind.java:
```diff
@@ -0,0 +1,113 @@
+package io.quarkus.datasource.common.runtime;
```
Mildly curious, this class has to be runtime? vs deployment time?

gsmet replied:
I want to be able to reference it everywhere. Typically it's used in the qualifier of the JDBC driver specific CDI beans.

emmanuelbernard commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/dialect/QuarkusPostgreSQL10Dialect.java:
```diff
@@ -0,0 +1,27 @@
+ * This is PostgreSQL's behavior.
+ * See https://github.com/quarkusio/quarkus/issues/1886
+ */
+public class QuarkusPostgreSQL10Dialect extends PostgreSQL95Dialect {
```
What's this one about? A move of class or a net new feature?

gsmet replied:
Ah silly me, it's supposed to extend `PostgreSQL10Dialect`. Let me fix that.

gsmet replied:
Fixed.

emmanuelbernard commented on extensions/jdbc/jdbc-mariadb/runtime/src/main/java/io/quarkus/jdbc/mariadb/runtime/MariaDBAgroalConnectionConfigurer.java:
```diff
@@ -0,0 +1,15 @@
+package io.quarkus.jdbc.mariadb.runtime;
```
runtime? why not deployment time?

gsmet replied:
Because it's used at runtime to configure the Agroal pool.

It's a CDI bean.

emmanuelbernard reviewed (dismissed):
Looks fine overall.

I had some general questions on why some classes where runtime and a few other points inline.
One I am specifically sensitive on is the `kind` name , which is too vague to me.
I also don't quite get on TLS is enable / disabled. I did not find a property.

emmanuelbernard commented:
BTW I had a OOME when running quarkus build (in the JPA - Infinispan Cache), is there a minimal default to enforce?

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AggregatedDataSourceBuildTimeConfigBuildItem.java:
```diff
@@ -0,0 +1,45 @@
+import io.quarkus.datasource.common.runtime.DataSourceUtil;
+import io.quarkus.datasource.runtime.DataSourceBuildTimeConfig;
+
+final class AggregatedDataSourceBuildTimeConfigBuildItem extends MultiBuildItem {
```
It aggregates all the different build time configuration objects into one aggregated and consistent view.

emmanuelbernard replied:
so aggregated as opposed ton JDBC and Reactive?

gsmet replied:
No. Aggregated as opposed to split into several smaller pieces you need to assemble. Not sure if I understand your concern?

gastaldi replied:
Maybe `Combined` would be a better term? Like in `CombinedIndexBuildItem`?

gsmet replied:
Well, aggregating the configuration is exactly what is is so I prefer keeping the name.

gsmet committed:
Add a test for unknown kind support in Agroal

gsmet commented:
@Sanne I just added a test: you just define a random kind and define the driver yourself.

gsmet commented:
Please don't merge, there's something weird with GH not picking up the latest version of my branch.

gsmet committed:
Rename quarkus.datasource.kind to quarkus.datasource.db-kind

aguibert commented on extensions/flyway/deployment/src/test/resources/baseline-on-migrate.properties:
```diff
@@ -1,7 +1,7 @@
-quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:test-quarkus-baseline-on-migrate;DB_CLOSE_DELAY=-1;INIT=RUNSCRIPT FROM 'src/test/resources/h2-init-data.sql'
-quarkus.datasource.driver=org.h2.Driver
+quarkus.datasource.db-kind=h2
```
Can we make `db-kind` optional for cases where there is only one DB kind enabled in the pom.xml?
This will be the case for probably >80% of apps, and it would reduce the number of required config options from 4 to 3.

aguibert replied:
Btw, my suggestion is not a hard requirement for this PR. The `db-kind` property could always be made optional later.

gsmet replied:
So the thing is that it's the only mandatory build time property so basically the only thing saying you have a datasource that is available at build time.

So for now, no.

emmanuelbernard replied:
Well we have the fact that a driver extension is present (and only one) via possible BuildItem publications. It's equally build time. But that shoudl have its own GitHub issue.

gsmet replied:
You know you have a driver extension but you don't know a datasource has been configured if all the datasource configuration is runtime as you shouldn't access the runtime config at build time.

We could try to assert some runtime configuration is here but that's very fragile. That's the main issue with doing this.

If we decide to do that, that would mean move more things at runtime as we wouldn't be able to know if a datasource is configured at build time.

emmanuelbernard replied:
Ok right, that's the core of the problem we can discuss on the ML but after this PR. 
Is it that bad to read runtime properties at build time and assume that some of its content is frozen at build time. In the JDBC url case, that would mean `jdbc:db-kind`

gastaldi commented on extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/UnknownDriverConfigTest.java:
```diff
@@ -0,0 +1,63 @@
+        assertTrue(dataSource.getConfiguration().metricsEnabled());
+        assertEquals(newConnectionSql, agroalConnectionFactoryConfiguration.initialSql());
+        try (Connection connection = dataSource.getConnection()) {
+        }
```
```suggestion
        assertThatCode(()->dataSource.getConnection().close()).doesNotThrowAnyException());
```

gsmet replied:
It's a copy of an existing test. I will leave it at that.

gastaldi commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcRuntimeConfig.java:
```diff
@@ -8,100 +8,71 @@
      */
     @ConfigItem
-    public Optional<Integer> initialSize;
+    public Optional<Integer> initialSize = Optional.empty();
```
Why is this necessary? I thought these fields would be set when injected?

gsmet replied:
Already explained above.

dmlloyd replied:
It should be an `OptionalInt` at any rate.

gastaldi commented on extensions/agroal/spi/src/main/java/io/quarkus/agroal/deployment/JdbcDriverBuildItem.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final Optional<String> xaDriverClass;
+
+    public JdbcDriverBuildItem(String kind, String driverClass, Optional<String> xaDriverClass) {
```
I can't help but think that the last parameter can be sent as null by other extensions. In order to prevent that, I'd suggest it to be String too and in the constructor body just use an `Optional.ofNullable`

gsmet replied:
I know it's not a pattern people love but I want to be clear it's not mandatory.

gsmet replied:
In the end I decided to expose two different constructors.

geoand commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalRecorder.java:
```diff
@@ -1,34 +1,27 @@
 @Recorder
 public class AgroalRecorder {
 
     public static final String DEFAULT_DATASOURCE_NAME = "<default>";
```
This is now longer used anymore as far as I can tell

gsmet replied:
Right, removed.

geoand replied:
Thanks

geoand commented on extensions/elytron-security-jdbc/deployment/src/main/java/io/quarkus/elytron/security/jdbc/deployment/ElytronSecurityJdbcProcessor.java:
```diff
@@ -50,7 +50,7 @@ FeatureBuildItem feature() {
             BuildProducer<SecurityRealmBuildItem> securityRealm,
             BeanContainerBuildItem beanContainerBuildItem, //we need this to make sure ArC is initialized
-            Optional<DataSourceInitializedBuildItem> dataSourceInitialized) throws Exception {
+            List<JdbcDataSourceBuildItem> dataSourcesConfigured) throws Exception {
```
This also requires the Javadoc to be updated

gsmet replied:
Done.

geoand reviewed (approved):
I added a couple minor comments. From my PoV, the PR is great!

geoand commented:
We will however need the tests to pass :)

```
2020-02-27T19:35:34.0133881Z [INFO] Running io.quarkus.it.main.DatasourceTestCase
2020-02-27T19:35:34.3687661Z [INFO] H2 database started in TCP server mode; server status: TCP server running at tcp://10.1.0.4:9092 (only local connections)
2020-02-27T19:35:38.0915387Z [INFO] H2 database was shut down; server status: Not started
2020-02-27T19:35:38.1561781Z [ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 4.125 s <<< FAILURE! - in io.quarkus.it.main.DatasourceTestCase
2020-02-27T19:35:38.1562445Z [ERROR] testDataSourceTransactions  Time elapsed: 0.015 s  <<< ERROR!
2020-02-27T19:35:38.1563666Z java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.util.NoSuchElementException: Property datasource.db-kind not found
```

is a failure I see in CI

gsmet committed:
Remove a now unused constant

gastaldi commented:
Too bad. At least it would assert something :(

Em Qui, 27 de fev de 2020 19:56, Guillaume Smet <notifications@github.com>
escreveu:

> *@gsmet* commented on this pull request.
> ------------------------------
>
> In
> extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/UnknownDriverConfigTest.java
> <https://github.com/quarkusio/quarkus/pull/7453#discussion_r385419208>:
>
> > +        try (Connection connection = dataSource.getConnection()) {
> +        }
>
> It's a copy of an existing test. I will leave it at that.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/7453?email_source=notifications&email_token=AAANG5IRN5G4CXOM33C5SLTRFBALBA5CNFSM4K4YYVW2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXIZ2KQ#discussion_r385419208>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAANG5KF2EF5MN437RUF4KLRFBALBANCNFSM4K4YYVWQ>
> .
>

gsmet commented:
Well, feel free to create a PR once this is in.

gsmet commented:
@geoand I pushed an update.
@emmanuelbernard I think I addressed all your concerns.

geoand commented:
@gsmet half of the JVM tests got canceled. I assume you want to restart them?

emmanuelbernard commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/LegacyDataSourceJdbcBuildTimeConfig.java:
```diff
@@ -0,0 +1,35 @@
+public class LegacyDataSourceJdbcBuildTimeConfig {
+
+    /**
+     * @deprecated use quarkus.datasource.kind (or quarkus.datasource.jdbc.driver if you really need a specific JDBC driver).
```
old reference to kind

gsmet replied:
Fixed.

emmanuelbernard commented on extensions/agroal/spi/src/main/java/io/quarkus/agroal/deployment/JdbcDataSourceBuildItem.java:
```diff
@@ -0,0 +1,36 @@
+
+    public JdbcDataSourceBuildItem(String name, String kind, boolean isDefault) {
+        this.name = name;
+        this.kind = kind;
```
not mandatory but renaming dbKind will help a future code reader

gsmet replied:
Done.

emmanuelbernard commented on extensions/agroal/spi/src/main/java/io/quarkus/agroal/deployment/JdbcDriverBuildItem.java:
```diff
@@ -0,0 +1,44 @@
+    private final Optional<String> xaDriverClass;
+
+    public JdbcDriverBuildItem(String kind, String driverClass, String xaDriverClass) {
+        this.kind = kind;
```
not mandatory but a future code reader will be helped by it being name dbKind

gsmet replied:
Done.

emmanuelbernard commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -710,32 +716,34 @@ private void handleHibernateORMWithNoPersistenceXml(
-        String error = driver.isPresent()
-                ? "Hibernate extension could not guess the dialect from the driver '" + resolvedDriver
+        String error = kind.isPresent()
+                ? "Hibernate extension could not guess the dialect from the kind '" + resolvedKind
```
```suggestion
                ? "Hibernate extension could not guess the dialect from the database kind '" + resolvedKind
```

gsmet replied:
Done and fixed a few other things in this message too.

gsmet commented:
The build failed due to the OIDC flow :/. Damn flaky tests. Will retry with the fix from Sergey.

gsmet commented:
Running the tests locally, will push soon a new update.

gsmet committed:
Make datasource.initial-size an OptionalInt

gsmet commented:
I addressed all the comments and CI is green except for an unrelated Windows build. I'm merging. We can iterate more in subsequent PRs if needed.

Thanks for the reviews!
